content/blog/2023-06-26-mismatch.Rmd

```diff
@@ -0,0 +1,85 @@
+
+- Construct the parameters for an API call based on the labels of our .csv on ArchiveDiffer.
+
+- Compare the data pulled from ArchiveDiffer with data from our API.
+
+The widget eventually outputs a csv file. Each row describes in detail the comparison between data in each file on ArchiveDiffer and data returned from our API given the same params. This comparison includes whether there is a difference between data from the two sources, and if there is, how many rows are impacted. 
+
+Additionally, the widget will identify if a whole file from ArchiveDiffer is missing. This happens when the API does not return any data given matching params from an ArchiveDiffer csv file. 
+
+
+## The Result
+There exists significant mismatches between data obtained from the ArchiveDiffer cache and covidcast API. 
+
+| Sources            | Number of files with row mismatches | Number of files in ArchiveDiffer | (%) |
+| :--------------------- | :--------: | :--------: | :------: |
+| jhu-csse          |   17165   | 92608 | 18.535116 |
+| chng              |   15725   | 39961 | 39.350867 |
+| usa-facts         |  9413   | 90392 | 10.413532 |
+| quidel            |  6055   | 89982 | 6.729124 |
+| dsew-cpr          |   2692   | 15652 | 17.199080 |
+| hhs               |   1506   | 44748 | 3.365514 |
+| indicator-combination|   288   | 288 | 100.000000 |
+| covid-act-now     |   280   | 7704 | 3.634476 |
+
+From this table, we see that source JHU has the highest number of of files with some row mismatches, closely followed by Change Healthcare. However, since JHU has a much bigger number of files on ArchiveDiffer in total, Change data from the API is actually the more inaccurate source by comparison (39% files have mismatches). The most inaccurate source when comparing API data to ArchiveDiffer data is indicator-combination. All indicator-combination data files on ArchiveDiffer currently have one or more rows different from the data returned from covidcast API.
+
+Full-file mismatch happens when the API returns no row at all when looking for data from a file on ArchiveDiffer. Fortunately, USAFacts is the only source with significant number of files where the full file is missing. 
+
+| Sources              | Number of full data missing from API |
+| :---------------- | :------: |
+| usa-facts        |   340   |
+| hhs           |   6   |
+| jhu-csse    |  5   |
+| dsew-cpr |  2   |
+| chng |  1   |
+
+Analysis by date shows that time in mismatched files are pretty evenly distributed. Overall, there is no particular period of time with less/more mismatch than others. 
+
+![](/blog/2023-06-26-mismatch/plot.png "Number of files on ArchiveDiffer with mismatch rows")
+
+## Widget Limitations
+This widget cannot tell if a mismatched row exists in S3 but not API or vice versa. It can only show the number of rows that should exists in both storage places but in reality, only exists in one of them.
```

Comment:
Other limitations:

```
For simplicity and expediency, this widget checks only the most recent version of the data, and ignores all data versioning information. Data versioning support in the Epidata API exists because it matters for modeling purposes _when_ information became available (this is the `issue` parameter in the covidcast endpoint). When we make repairs, Epidata permits us to correct old issues, but S3 does not permit us to correct old uploads. This means that patching typically only touches the API, and leaves ArchiveDiffer data alone even if the data being patched is the most recent version available. This widget does not check the timestamp of S3 uploads for ArchiveDiffer data or the `issue` column of API data. This means that it includes differences due to patches, which we might wish to ignore.
```